File: cli/internal/submit/uid.go
package submit

import (
	"crypto/rand"
	"fmt"
)

const crockfordAlphabet = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"

// EncodeCrockford encodes a byte slice to Crockford base32.
// Uses bit-buffer algorithm: accumulates 8 bits per byte, extracts 5-bit groups MSB-first.
func EncodeCrockford(data []byte) string {
	var bits, value int
	var output []byte

	for _, b := range data {
		value = (value << 8) | int(b)
		bits += 8
		for bits >= 5 {
			bits -= 5
			output = append(output, crockfordAlphabet[(value>>bits)&0x1f])
		}
	}
	if bits > 0 {
		output = append(output, crockfordAlphabet[(value<<(5-bits))&0x1f])
	}
	return string(output)
}

// GenerateUID returns a new 16-character Crockford base32 UID using 10 random bytes.
func GenerateUID() (string, error) {
	buf := make([]byte, 10)
	if _, err := rand.Read(buf); err != nil {
		return "", fmt.Errorf("uid: cannot read random bytes: %w", err)
	}
	return EncodeCrockford(buf), nil
}

File: cdk/lambda/shared/uid.ts
import { randomBytes } from "crypto";

const CROCKFORD_ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ";

/**
 * Encode a Buffer to Crockford base32 using a bit-buffer algorithm.
 * Reads 8 bits per byte, extracts 5-bit groups MSB-first via unsigned right shift.
 */
export function encodeCrockford(buf: Buffer): string {
  let bits = 0;
  let value = 0;
  let output = "";

  for (const byte of buf) {
    value = (value << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      output += CROCKFORD_ALPHABET[(value >>> bits) & 0x1f];
    }
  }

  if (bits > 0) {
    output += CROCKFORD_ALPHABET[(value << (5 - bits)) & 0x1f];
  }

  return output;
}

/**
 * Generate a 16-character Crockford base32 UID from 10 random bytes (80 bits entropy).
 */
export function generateUid(): string {
  return encodeCrockford(randomBytes(10));
}
